providers/implementations/signature/rsa.c

```diff
@@ -0,0 +1,922 @@
+        char mdprops[80] = "";
+
+        str = mdname;
+        if (!OSSL_PARAM_get_utf8_string(p, &str, sizeof(mdname)))
+            return 0;
+
+        str = mdprops;
+        if (propsp != NULL
+            && !OSSL_PARAM_get_utf8_string(propsp, &str, sizeof(mdprops)))
+            return 0;
+
+        /* For checking purposes */
+        md = EVP_MD_fetch(prsactx->libctx, mdname, mdprops);
+
+        if (rsa_pss_restricted(prsactx)) {
+            if (md != NULL && !EVP_MD_is_a(prsactx->md, mdname)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_DIGEST_NOT_ALLOWED);
+                ret = 0;
+            }
+        }
+        if (ret && !check_padding_md(md, prsactx->pad_mode))
+            ret = 0;
+        EVP_MD_free(md);
+        if (!ret)
+            return 0;
+
+        OPENSSL_strlcpy(prsactx->mdname, mdname, sizeof(prsactx->mdname));
+        OPENSSL_strlcpy(prsactx->mdprops, mdprops, sizeof(prsactx->mdprops));
+        EVP_MD_free(prsactx->md);
+        prsactx->md = NULL;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_DIGEST_SIZE);
+    if (p != NULL && !OSSL_PARAM_get_size_t(p, &prsactx->mdsize))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_PAD_MODE);
+    if (p != NULL) {
+        int pad_mode = RSA_NO_PADDING;
+
+        if (!OSSL_PARAM_get_int(p, &pad_mode))
+            return 0;
+        switch (pad_mode) {
+        case RSA_PKCS1_OAEP_PADDING:
+            /*
+             * OAEP padding is for asymmetric cipher only so is not compatible
+             * with signature use.
+             */
+            ERR_raise(ERR_LIB_RSA,
+                      PROV_R_ILLEGAL_OR_UNSUPPORTED_PADDING_MODE);
+            return 0;
+        case RSA_PKCS1_PSS_PADDING:
+            if (prsactx->mdname[0] == '\0') {
+                OPENSSL_strlcpy(prsactx->mdname, "SHA1",
+                                sizeof(prsactx->mdname));
+                OPENSSL_strlcpy(prsactx->mdprops, "",
+                                sizeof(prsactx->mdprops));
+                EVP_MD_free(prsactx->md);
+                prsactx->md = NULL;
+            }
+            goto cont;
+        case RSA_PKCS1_PADDING:
+        case RSA_SSLV23_PADDING:
+        case RSA_NO_PADDING:
+        case RSA_X931_PADDING:
+            if (RSA_get0_pss_params(prsactx->rsa) != NULL) {
+                ERR_raise(ERR_LIB_RSA,
+                          PROV_R_ILLEGAL_OR_UNSUPPORTED_PADDING_MODE);
+                return 0;
+            }
+        cont:
+            if (!check_padding_md(prsactx->md, pad_mode))
+                return 0;
+            break;
+        default:
+            return 0;
+        }
+        prsactx->pad_mode = pad_mode;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_PSS_SALTLEN);
+    if (p != NULL) {
+        int saltlen;
+
+        if (prsactx->pad_mode != RSA_PKCS1_PSS_PADDING) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_PSS_SALTLEN);
```

Comment:
Should the reason value be different here?

Reply:
Why?  This code handle the OSSL_SIGNATURE_PARAM_PSS_SALTLEN param, and *any* saltlen setting is invalid unless in PSS padding mode

Reply:
If you want, I could make that an `ERR_raise_data()` and have additional text explaining this further.

Reply:
I'd see something like PROV_R_NOT_SUPPORTED or PROV_R_INVALID_PARAMETER to be more appropriate because the PROV_R_INVALID_PSS_SALTLEN would mean to me that the value is incorrect not that even the appearance of the saltlen param is incorrect.

Reply:
I opted for already existing PROV_R_NOT_SUPPORTED, and added additional text.